proposals/NNNN-library-evolution.md

```diff
@@ -0,0 +1,374 @@
+
+This proposal introduces:
+
+- a "library evolution" build mode for libraries that are declaring ABI stability, which preserves the ability to make certain changes to types without breaking the library's ABI; and
+- an attribute for such libraries to opt out of this flexibility on a per-type basis, allowing certain compile-time optimizations.
+
+The mechanisms for this are already in place, and were used to stabilize the
+ABI of the standard library. This proposal makes them features for use by
+any 3rd-party library that wishes to declare itself ABI stable.
+
+**This build mode will have no impact on libraries built and distributed with an app**. Such libraries will receive the same optimization that they have in previous versions of Swift.
+
+_Note: this proposal will use the word "field" to mean "stored instance property"._
+
+## Motivation
+
+Starting in Swift 5, libraries are able to declare a stable ABI, allowing a  library binary to be replaced with a newer version without requiring client programs to be recompiled. 
+
+What consitutes the ABI of a library differs from language to language. In C and C++, the public ABI for a library includes information that ideally would be kept purely as an implementation detail. For example, the _size_ of a struct is fixed as part of the ABI, and is known to the library user at compile time. This prevents adding new fields to that type in later releases once the ABI is declared stable. If direct access to fields is allowed, the _layout_ of the struct can also become part of the ABI, so fields cannot then be reordered.
+
+This often leads to manual workarounds. A common technique is to have the struct hold only a pointer to an "impl" type, which holds the actual stored properties. All access to these properties is made via function calls, which can be updated to handle changes to the layout of the "impl" type. This has some obvious downsides:
+
+- it introduces indirection and function call overhead when accessing fields;
+- it means the fields of the struct must be stored on the heap even when they could otherwise be stored on the stack, with that heap allocation needing to be managed via reference counting or some other mechanism;
+- it means that the fields of the struct data are not stored in contiguous memory when placed in an array; and 
+- the implementation of all this must be provided by the library author.
+
+A similar challenge occurs with enums. As discussed in [SE-0192][], introducing new cases to an enum can break source compatability. There are also ABI consequences to this: adding new enum cases sometimes means increasing the storage needed for the enum, which can affect the size and layout of an enum in the same way adding fields to a struct can.
```

Comment:
Typo: "compatability". Also, you may want to distinguish the C and Swift cases here. In C (well, Objective-C) this isn't a problem because you specify an underlying type.